### PR TITLE
Add rock paper scissors game example

### PR DESCRIPTION
**Pull request by xermicus**
While warming up with the stack I created a little rock paper scissors game smart contract. I'm polishing it now and would appreciate some early feedback :blush: 

This is not done yet, I want to add extensive testing, more comments explaining how it works and possible a little testing client.

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
## 🦑 📈 ink! Example Contracts ‒ Changes Report 📉 🦑
 These are the results when building the `examples/*` contracts from this branch with `cargo-contract 1.4.0-25bfb96` and comparing them to ink! `master`: 

|                          | Δ Optimized Size | Δ Used Gas | Total Optimized Size | Total Used Gas |
|:--------------------------|------------------:|------------:|----------------------:|----------------:|
| `accumulator`            |                  |            | 1.00 K               |                |
| `adder`                  |                  |            | 2.04 K               |                |
| `contract-introspection` |                  |            | 2.32 K               |                |
| `contract-terminate`     |                  |            | 0.92 K               | 275_000        |
| `contract-transfer`      |                  |            | 8.31 K               | 75_000         |
| `data-structures`        |                  |            | 1.73 K               |                |
| `delegate-calls`         |                  |            | 2.89 K               | 76_242         |
| `delegator`              |                  |            | 6.34 K               | 232_135        |
| `dns`                    |                  |            | 8.81 K               | 225_000        |
| `erc1155`                |                  |            | 16.74 K              | 450_000        |
| `erc20`                  |                  |            | 8.42 K               | 225_000        |
| `erc721`                 |                  |            | 11.62 K              | 600_000        |
| `flipper`                |                  |            | 1.24 K               | 75_000         |
| `forward-calls`          |                  |            | 2.87 K               | 151_411        |
| `incrementer`            |                  |            | 1.14 K               |                |
| `mother`                 |                  |            | 12.16 K              |                |
| `multisig`               |                  |            | 25.28 K              | 470_240        |
| `payment-channel`        |                  |            | 7.95 K               |                |
| `rand-extension`         |                  |            | 3.79 K               | 75_000         |
| `rock-paper-scissors`    | +17.54 K         |            | 17.54 K              |                |
| `seal-code-hash`         |                  |            | 1.39 K               |                |
| `seal-ecdsa`             |                  |            | 1.71 K               |                |
| `set-code-hash`          |                  |            | 1.49 K               | 150_000        |
| `subber`                 |                  |            | 2.06 K               |                |
| `trait-erc20`            |                  |            | 8.69 K               | 225_000        |
| `trait-flipper`          |                  |            | 0.97 K               | 75_000         |
| `trait-incrementer`      |                  |            | 1.12 K               | 150_000        |
| `updated-incrementer`    |                  |            | 9.72 K               |                |
| `upgradeable-flipper`    |                  |            | 1.48 K               |                |



[Link to the run](https://gitlab.parity.io/parity/ink-waterfall/-/pipelines/208371) | Last update: Tue Aug 16 12:50:53 CEST 2022

File: examples/rock-paper-scissors/game.rs [hunk 1]
```diff
@@ -0,0 +1,494 @@
...
+};
+use scale::{
+    Decode,
+    Encode,
+};
+
+type BlockNo = u32;
+
+/// How many blocks may pass until the game is considered to be timed out.
+pub(crate) const TIMEOUT: BlockNo = 100;
```
Comment (cmichi) [inline examples/rock-paper-scissors/game.rs:25]: Please use `BlockNumber` here, see e.g. https://github.com/paritytech/ink/blob/master/examples/mother/lib.rs#L72. You'll have to move those definitions inside the `#[ink::contract]` macro for it to work.

The context here is that the block number is a generic type in Substrate. So could be anything. By using ink!'s `BlockNumber` you'll use the chain's type.

File: examples/rock-paper-scissors/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,35 @@
+[package]
+name = "rock_paper_scissors"
+version = "0.1.0"
+authors = ["[your_name] <[your_email]>"]
+edition = "2021"
+
+[dependencies]
+ink_primitives = { version = "3", default-features = false }
```
Comment (cmichi) [inline examples/rock-paper-scissors/Cargo.toml:8]: Could you merge `master` and upgrade to `4.0.0-alpha.1`?

File: examples/rock-paper-scissors/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,35 @@
+[package]
+name = "rock_paper_scissors"
+version = "0.1.0"
+authors = ["[your_name] <[your_email]>"]
```
Comment (cmichi) [inline examples/rock-paper-scissors/Cargo.toml:4]:
```suggestion
authors = ["Parity Technologies <admin@parity.io>"]
```

**Comment by cmichi** [comment]
Can you access the logs of the failing CI stages on GitLab? There are some `clippy` suggestions and `spellcheck` also got lucky ;-).

**Comment by xermicus** [comment]
Sorry @cmichi this is on my back log (thanks for the feedback)! Would be nice to have this example in the ink v4 release :blush:

**Comment by HCastano** [comment]
Gonna close this, free free to re-open if you start working on it again.

**Pull request closed without merging**